stdlib/src/collections/string.mojo

```diff
@@ -1340,6 +1343,15 @@ struct String(
         Returns:
             The joined string.
         """
+
+        # TODO(#3403): the renaming of the following method from join to fast_join and
+        # the _type_is_eq() call is a hack needed to avoid the following error:
+        # "ambiguous call to 'join', each candidate requires 0 implicit conversions,
+        # disambiguate with an explicit cast"
+        @parameter
+        if _type_is_eq[T, String]() or _type_is_eq[T, StringLiteral]():
+            return self.fast_join(rebind[List[String]](elems))
```

Comment:
This will not compile for `StringLiteral` since rebind is not supposed to use the constructors AFAIK, and if it does, this will alloc unnecessarily.
```suggestion
        if _type_is_eq[T, String]():
            return self.fast_join(rebind[List[String]](elems))
        elif _type_is_eq[T, StringLiteral]():
            return self.fast_join(rebind[List[StringLiteral]](elems))
        # elif _type_is_eq[T, StringSlice]() # FIXME(#3597): once StringSlice conforms to CollectionElement trait
        #     return self.fast_join(rebind[List[StringSlice]](elems))
```

Reply:
Done here: https://github.com/msaelices/mojo/commit/907c7865d1cca2dc9af04869b0f91d2538a58ebf

The previous code did compile. Even with your suggestion to save one allocation, it was not as easy as I needed to overload the `as_bytes_span()` method to match the `SizedSpanableCollectionElement` trait signature. 

I'm not entirely sure if the optimization is so good as we usually don't deal with thousands of handwritten string literals, and it's more effective to optimize the case of reading strings from a CSV file, for example.

But all the optimizations are welcome, so thanks for the tip.

Reply:
> it's more effective to optimize the case of reading strings from a CSV file, for example

That is where the `elif _type_is_eq[T, StringSlice]()` comes in, once #3597 lands I'm planning to implement `.split()` and refactor `splitlines()` for `StringSlice` so that they return `List[StringSlice]`. And mixing that with when we have some memory mapping or for now just building a span out of the list that `file.read_bytes()` returns, we will get some crazy performance improvements higher up the stack :D . 

Especially for csv parsing if after separating with `split(",")` and `splitlines()` you want to do some transformations to the data and copy it elsewhere or build a view to the raw bytes using some zero-copy data structure. Then if you wanna do some transformations to it and join it as a csv again with you don't need to alloc too many Strings for all that :rocket: . And this join logic you built here also applies to anything to do with two buffers with data you want to join, we'll just have to tweak it a bit to make it more generic if we find a reason for it :fire:  .

> But all the optimizations are welcome, so thanks for the tip.

Yeah these kind of optimizations add up, we'll see the fruits of our labor once online "fastest language" videos start mentioning Mojo ;)